cs/src/core/ClientSession/AdvancedClientSession.cs

```diff
@@ -763,14 +836,35 @@ public int GetLength(ref Value t, ref Input input)
                 return _clientSession.variableLengthStruct.GetLength(ref t, ref input);
             }
 
-            public void InitialUpdater(ref Key key, ref Input input, ref Value value, long address)
+            public void InitialUpdater(ref Key key, ref Input input, ref Value value)
             {
-                _clientSession.functions.InitialUpdater(ref key, ref input, ref value, address);
+                _clientSession.functions.InitialUpdater(ref key, ref input, ref value);
             }
 
-            public bool InPlaceUpdater(ref Key key, ref Input input, ref Value value, long address)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool InPlaceUpdater(ref Key key, ref Input input, ref Value value, ref RecordInfo recordInfo, long address)
+                => !this.SupportsLocks
+                    ? InPlaceUpdaterNoLock(ref key, ref input, ref value, ref recordInfo, address)
+                    : InPlaceUpdaterLock(ref key, ref input, ref value, ref recordInfo, address);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private bool InPlaceUpdaterNoLock(ref Key key, ref Input input, ref Value value, ref RecordInfo recordInfo, long address)
+                => _clientSession.functions.InPlaceUpdater(ref key, ref input, ref value, ref recordInfo, address)
+                    && _clientSession.fht.UpdateSIForIPU(ref value, address);
+
+            private bool InPlaceUpdaterLock(ref Key key, ref Input input, ref Value value, ref RecordInfo recordInfo, long address)
             {
-                return _clientSession.functions.InPlaceUpdater(ref key, ref input, ref value, address);
+                long context = 0;
+                this.Lock(ref recordInfo, ref key, ref value, OperationType.RMW, ref context);
```

Comment:
It seems overkill to send OperationType to Lock. All we need to know here is whether this is a reader lock or writer lock. That is why I suggested either a new enum `OpType` or just a bool.

Reply:
Maybe call it LockType.Shared and Exclusive?